### PR TITLE
planner: PhysicalPlan memory trace 2

### PR DESCRIPTION
**Pull request by fzzf678**
### What problem does this PR solve?

Issue Number: ref #37632 

Problem Summary:

### What is changed and how it works?
Add MemoryUsage() for
* CIStr
* Range
* PhysicalUnionScan
* PhysicalIndexScan
* PhysicalTableScan

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: util/ranger/types.go [hunk 1]
```diff
@@ -247,3 +248,23 @@ func formatDatum(d types.Datum, isLeftSide bool) string {
...
+const emptyRangeSize = int64(unsafe.Sizeof(Range{}))
+
+// MemoryUsage return the memory usage of Range
+func (ran *Range) MemoryUsage() (sum int64) {
+	if ran == nil {
+		return
+	}
+
+	sum = emptyRangeSize +
+		int64(cap(ran.LowVal)-len(ran.LowVal))*int64(unsafe.Sizeof(types.Datum{})) +
```
Comment (qw4990) [inline util/ranger/types.go:261]: We can just use `EmptyDatumSize` here instead of `SizeOf(Datum{})`.

File: util/ranger/types.go [hunk 2]
```diff
@@ -247,3 +248,23 @@ func formatDatum(d types.Datum, isLeftSide bool) string {
...
+		int64(cap(ran.HighVal)-len(ran.HighVal))*int64(unsafe.Sizeof(types.Datum{}))
```
Comment (qw4990) [inline util/ranger/types.go:262]: It seems a little complicated, how about removing the last two parts?
Just use the formula `emptyRangeSize + LowVals.MemUsage() + HighVals.MemUsage()`.

File: planner/core/physical_plans.go [hunk 3]
```diff
@@ -131,6 +131,27 @@ type PartitionInfo struct {
...
+	for _, cond := range pi.PruningConds {
+		sum += cond.MemoryUsage()
+	}
+	for _, cis := range pi.PartitionNames {
+		sum += cis.MemoryUsage()
+	}
+	for _, col := range pi.Columns {
+		sum += col.MemoryUsage()
+	}
+	return
```
Comment (qw4990) [inline planner/core/physical_plans.go:152]: Should we accumulate memory usage of `ColumnNames` here?

Comment (fzzf678) [reply planner/core/physical_plans.go:152]: > Should we accumulate memory usage of `ColumnNames` here?

yes, you are right

**Review by Reminiscent — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**